zilliqa/src/exec.rs

```diff
@@ -497,9 +496,66 @@ impl State {
             })
             .build();
 
-        let e = evm.transact()?;
+        let transact_result = evm.transact();
+
         let (mut state, cfg) = evm.into_db_and_env_with_handler_cfg();
-        Ok((e, state.finalize(), cfg.env))
+
+        let result_and_state = match transact_result {
+            Err(
+                EVMError::Transaction(_)
+                | EVMError::Header(_)
+                | EVMError::Custom(_)
+                | EVMError::Precompile(_),
+            ) => {
+                let account = state.load_account(from_addr)?;
+                let Account {
+                    code,
+                    nonce,
+                    balance,
+                    ..
+                } = &mut account.account;
+
+                let code_hash = match &code {
+                    EvmCode(vec) if vec.is_empty() => KECCAK_EMPTY,
+                    // Just signal that the code exists by returning non-KECCAK_EMPTY hash (we don't have to be precise here)
+                    EvmCode(_) => B256::ZERO,
+                    _ => KECCAK_EMPTY,
+                };
+                let code = match code {
+                    EvmCode(code) => mem::take(code),
+                    _ => bail!("Evm account contains non-evm code type!"),
+                };
+
+                // The only updated fields are nonce and balance, rest can be set to default (means: no update)
+                let affected_acc = revm::primitives::Account {
```

Comment:
This feels like we are re-implementing things that should really be handled inside `revm`. I'm quite worried about doing this because when we did it before (when we depended on sputnik) we got lots of the edge cases wrong.

Just to summarise my understanding of the root issue here - The problem is that transactions get dropped from the transaction pool if we try to execute them and after executing, realise that there is not enough space left in the block. Could we not instead solve that issue by being more strict on the condition by which we consider executing a transaction - Instead of (optimistically) executing the next transaction and checking it fits, what if we checked the maximum gas spend of a transaction before executing it and stop building the block if it doesn't fit. That way, we only execute a transaction once we are certain it can be included in the block. `reth` does this and I don't think its a problem from the user's perspective if their transaction with an overly large gas limit gets included one block later than it theoretically could have.

Am I making sense? Happy to have a chat to discuss this instead if you prefer.

cc @DrZoltanFazekas

Reply:
I preproduced the case on the Holesky testnet, which we can consider a "reference" of correct behaviour. First I reduced the balance of my account `0xd819ffce7a58b1e835c25617db7b46a00888b013` to `0.000008323450116360 ether`. Then I sent 10 simple transfer transactions knowing that after 5 of them my account will run out of gas. The first 5 transactions were mined in the same block, you can see them as the 5 topmost transactions in the list: https://holesky.etherscan.io/address/0xd819ffce7a58b1e835c25617db7b46a00888b013. I also got the receipts for those 5 transactions. Then my script got stuck waiting for further receipts and timed out minutes later with:
```
AggregateError: 
    at internalConnectMultiple (node:net:1118:18)
    at internalConnectMultiple (node:net:1186:5)
    at Timeout.internalConnectMultipleTimeout (node:net:1712:5)
    at listOnTimeout (node:internal/timers:583:11)
    at processTimers (node:internal/timers:519:7)
```

In the Holesky RPC node's mempool there was no `pending` or `queued` transaction from my account, so the other 5 transactions for which there wasn't enough balance left must have been dropped. 

Note the difference between Holesky and our prototestnet: on the prototestnet the validators would drop the other 5 transactions, but they would stay in the API nodes' mempool as `pending` (forever).